### PR TITLE
Glitch server

### PR DESCRIPTION
**Pull request by pangadev**
**Issues:**

After redemption, the sidebar won't update right away. It seems that it takes some time to route data between voucherify and demostore. If I would fetch customers list right after redemption, the list would still be the same as before (only after few refreshes it changes)

Refreshing on subpages results in fail (on localhost)

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,2 +1,4 @@
 PORT=3000
-NODE_ENV=12
\ No newline at end of file
+NODE_ENV=12
```
Comment (pawelrychlik) [inline .env:2]: ?

File: server.js [hunk 2]
```diff
@@ -39,12 +39,22 @@ app.get("/ping", (req, res) => {
...
-app.use(express.static("build"));
+const voucherify = voucherifyClient({
+  applicationId: process.env.APPLICATION_ID,
+  clientSecretKey: process.env.CLIENT_SECRET_KEY,
+});
+
+app.get("/customers", async (request, response) => {
+  try {
+    console.log("[Fetching customers]");
+    voucherify.customers.list().then((customers) => response.json(customers));
```
Comment (pawelrychlik) [inline server.js:50]:
```
const customers = await v.customers.list();
response.json(customers);
```

File: src/components/Cart/CartTotals.js [hunk 3]
```diff
@@ -51,10 +51,13 @@ export default function CartTotals({ value }) {
         >
           <CustomerConsumer>
             {(ctx) => {
+              let customer = ctx.customers.filter(
```
Comment (pawelrychlik) [inline src/components/Cart/CartTotals.js:54]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find
`ctx.customers.find(customer => customer.name === ctx.customer)`

btw don't you have `ctx.customer` available already?

Comment (pawelrychlik) [reply src/components/Cart/CartTotals.js:54]: this find/filter should not be done here, context should provide you the selected customer already

File: src/components/CustomerContext.js [hunk 4]
```diff
@@ -19,25 +19,38 @@ const loadItemsFromLocalStorage = () => {
...
+    this.setState(loadItemsFromLocalStorage());
+    this.getCustomers();
   }
 
+  getCustomers = async () => {
+    try {
+      let customers = await fetch("/customers").then((x) => x.json());
+      this.setState({
+        customers: customers.customers,
+      });
```
Comment (pawelrychlik) [inline src/components/CustomerContext.js:39]: formatting

```
let { customers } = await fetch("/customers").then((res) => res.json());
this.setState({
  customers,
});
```

Comment (pawelrychlik) [reply src/components/CustomerContext.js:39]: and I think you'll need this fetch option soon: `credentials: 'include'`

File: src/components/SidebarContent.js [hunk 5]
```diff
@@ -1,32 +1,38 @@
...
 import { CustomerConsumer } from "./CustomerContext";
 
 export default function SidebarContent() {
   return (
     <div className="side-menu-container">
       <CustomerConsumer>
         {(ctx) => {
+          let customer = ctx.customers.filter(
+            (customer) => customer.name === ctx.customer
+          )[0];
```
Comment (pawelrychlik) [inline src/components/SidebarContent.js:12]: see? you keep copy-pasting it all over the place

Comment (pawelrychlik) [reply src/components/SidebarContent.js:12]: the easiest (not the best from a tech point of view, though) would be to add a function: `ctx.getCustomer()` which does this: `this.state.customers.find((customer) => this.state.customer === customer.name)`

File: src/components/Cart/CartTotals.js (hunk 3 above)
Comment (pawelrychlik) [reply src/components/Cart/CartTotals.js:54]: plz see my other comment below

File: .env (hunk 1 above)
Comment (pangadev) [reply .env]: my mistake. I need to force using this app in node version 12, as on newer versions there is an issue with sql package.

Comment (pawelrychlik) [reply .env]: you can indicate node version in package.json: https://docs.npmjs.com/files/package.json#engines
but afaik it's just a suggestion, that might not be respected by some PaaS platforms.

NODE_ENV is typically either "development" or "production". This flag could be used to e.g. differentiate between our "npm start" and "npm run dev".

File: src/components/Sidebar/SidebarContent.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  return (
+    <div className="side-menu-container">
+      <CustomerConsumer>
+        {(ctx) => {
+          return (
+            <>
+              <select
+                id="storeCustomers"
+                onChange={(e) => ctx.setCustomer(e.target.value)}
+                value={(Object(ctx.customer) || {}).id || "DEFAULT"}
```
Comment (pawelrychlik) [inline src/components/Sidebar/SidebarContent.js:15]: what's the point of wrapping with Object() ?

File: src/components/Sidebar/SidebarContent.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                <option value="DEFAULT" disabled>
+                  Select customer
+                </option>
+                {ctx.customers.map((customer) => (
+                  <option key={customer.name} value={customer.id}>
+                    {customer.name}
+                  </option>
+                ))}
+              </select>
+              {ctx.fetchingCustomer ? (
```
Comment (pawelrychlik) [inline src/components/Sidebar/SidebarContent.js:26]: and, does the 5sec delay work? :)

File: src/components/Sidebar/SidebarContent.js (hunk 6 above)
Comment (pangadev) [reply src/components/Sidebar/SidebarContent.js:15]: My mistake, something wasn't working, but now it works.

File: src/components/Sidebar/SidebarContent.js (hunk 7 above)
Comment (pangadev) [reply src/components/Sidebar/SidebarContent.js:26]: I believe so.

**Pull request merged**